File: Frontend/src/app/classes/programacao/programacao.ts
import { Usina } from './../usina/usina';
import { Proposta } from './../proposta/proposta';
import { Endereco } from '../endereco/endereco';
import { ResistenciaTipo, Pedra, Slump, Uso } from '../traco/traco.classes';
import { Vendedor } from '../vendedor/vendedor';
import { ProgramacaoDemaisServicos } from './programacao-demais-servicos';

export class Status { codigo: number; descricao: string; color: string}

export enum EProgramacaoConfirmacao {
    Sim = 'S',
    Nao = 'N',
    Parcial = 'P'
}

export enum EProgramacaoTipoAlteracao {

    Insert = 1,
    Copy = 2,
    Update = 3

}

export const confirmacaoOpcoes: { codigo: EProgramacaoConfirmacao; descricao: string; }[] = [
    {codigo: EProgramacaoConfirmacao.Sim, descricao: "Sim" },
    {codigo: EProgramacaoConfirmacao.Nao, descricao: "Não" },
    {codigo: EProgramacaoConfirmacao.Parcial, descricao: "Parcial" }
];

export class CorpoDeProvaMoldador { codigo: number; descricao: string; }
export const corpoDeProvaMoldadores: CorpoDeProvaMoldador[] = [
    {codigo: 1, descricao: "MOTORISTA"},
    {codigo: 2, descricao: "LABORATÓRIO"}
];

export class CorpoDeProvaTipo { descricao: string; }
export const corpoDeProvaTipos: CorpoDeProvaTipo[] = [
    {descricao: "Nenhum"},
    {descricao: "Parcial"},
    {descricao: "Total"}
];

export class CorpoDeProvaMoldagemRemota { codigo: number; descricao: string; }
export const corpoDeProvaMoldagemRemota: CorpoDeProvaMoldagemRemota[] = [
    {codigo: 1, descricao: "Padrao da Central"},
    {codigo: 2, descricao: "Na usina"},
    {codigo: 3, descricao: "Na obra"}
];

export enum EProgramacaoStatus {
    AguardandoConfirmacao = 9161,
    Programado = 9162,
    Rejeitado = 9163,
    Cancelada = 9164,
    Revalidacao = 9165,
    AguardandoAnaliseLimiteCredito = 9166,
    LimiteCreditoInsuficiente = 9167,
    AprovacaoInadimplente = 9168
}

export const statusProgramacao: Status[] = [
    {codigo: EProgramacaoStatus.AguardandoConfirmacao , descricao: "Aguardando confirmação", color:'#ffc800' },
    {codigo: EProgramacaoStatus.Programado , descricao: "Programado", color:'#00a500' },
    {codigo: EProgramacaoStatus.Rejeitado , descricao: "Rejeitado", color:'#FF8072' },
    {codigo: EProgramacaoStatus.Cancelada , descricao: "Cancelada" , color:'#FF8072'},
    {codigo: EProgramacaoStatus.Revalidacao , descricao: "Revalidação de programação", color:'#ffc800' },
    {codigo: EProgramacaoStatus.AguardandoAnaliseLimiteCredito , descricao: "Aguardando Análise de Limite de Crédito", color: 'FF8072' },
    {codigo: EProgramacaoStatus.LimiteCreditoInsuficiente , descricao: "Limite de Crédito Insuficiente", color: '#FF8072' },
    {codigo: EProgramacaoStatus.AprovacaoInadimplente, descricao: "Aguardando Aprovação de Inadimplência.", color:'#FF8072' }
];

export class Programacao{
    usina: Usina;
    contratoAno: number = 0;
    contratoNumero: number = 0;
    sequencia: number = 0;
    propostaAno: number = 0;
    propostaNumero: number = 0;
    proposta: Proposta;
    obraNumero: number = 0;
    obraNome: string = '';
    usinaEntrega: Usina;
    endereco: Endereco;

    dataConcretagem: Date;
    horario: string = '';
    necessitaConfirmacao: EProgramacaoConfirmacao = EProgramacaoConfirmacao.Sim;
    volumeLiberado: number = 0;
    intervaloEmMinutosEntreCargas: number = 0;
    
    resistenciaTipo: ResistenciaTipo
    mpa: number = 0;
    consumo: number = 0;
    pedra: Pedra;
    slump: Slump;
    slumpNotaFiscal: string;
    uso: Uso;
    pecaConcretar: string;
    andar: string;
    obraTracoSequencia: number = 0;
    
    volumeTotal: number = 0;
    volumeEntregue: number = 0;
    volumePorCarga: number = 0;
    
    tracoPesadoResistenciaTipo: ResistenciaTipo;
    tracoPesadoMpa: number = 0;
    tracoPesadoConsumo: number = 0;
    tracoPesadoPedra: Pedra;
    tracoPesadoSlump: Slump;
    tracoPesadoUso: Uso;
    
    obraBombaSequencia: number = 0;
    distanciaTubulacao: number = 0;
    horarioBomba: string = '';

    demaisServicos: ProgramacaoDemaisServicos[] = [];

    vibradorQuantidade: number = 0;
    vibradorValorUnitario: number = 0;
    vibradorValorTotal: number = 0;
    vibradorVendedor: Vendedor;
    
    status: EProgramacaoStatus = EProgramacaoStatus.AguardandoConfirmacao;

    solicitante: string = '';
    observacao: string = '';

    observacaoInterna: string = '';

    corpoDeProvaTipo: string = 'Nenhum';
    corpoDeProvaQuantidade: number = 0;
    corpoDeProvaIntervalo: number = 0;
    corpoDeProvaMoldador: number = 0;
    corpoDeProvaMoldagemRemota: string = '';

    temNotaFicalEmitida: boolean = false;

    obraFrenteSequencia: number = 0;

    valorTotal: number = 0;
    valorConcreto: number = 0;
    valorExtras: number = 0;
    valorBomba: number = 0;
    valorDemaisServicos: number = 0;

    valorTotalRemessasEmitidas: number = 0;

    tempoAteAObra: number = 0;
    tempoBtNaObra: number = 0;
    tempoDescarga: number = 0;
}

File: Backend/TopSys.TopConWeb.Domain/Entities/Programacao.cs
using System;
using System.Collections.Generic;
using TopSys.TopConWeb.Domain.Enums;
using TopSys.TopConWeb.Domain.Interfaces.Entities;

namespace TopSys.TopConWeb.Domain.Entities
{
    public class Programacao : IHasEndereco
    {
        public int UsinaCodigo { get; set; }
        public virtual Usina Usina { get; set; }
        
        public int? ContratoAno { get; set; } = 0;
        public int? ContratoNumero { get; set; } = 0;
        public virtual Contrato Contrato { get; set; }

        public int Sequencia { get; set; }
        
        public int? PropostaAno { get; set; } = 0;
        public int? PropostaNumero { get; set; } = 0;
        public virtual Proposta Proposta { get; set; }

        public int ObraNumero { get; set; }
        
        public string ObraNome { get; set; }
        
        public int UsinaEntregaCodigo { get; set; }
        public virtual Usina UsinaEntrega { get; set; }
        
        public string EnderecoCep { get; set; }
        public string EnderecoLogradouro { get; set; }
        public int EnderecoNumero { get; set; }
        public string EnderecoComplemento { get; set; }
        public string EnderecoBairro { get; set; }
        public int? EnderecoMunicipioCodigo { get; set; } = 0;
        public virtual Municipio EnderecoMunicipio { get; set; }
        
        public int? ResistenciaTipoCodigo { get; set; } = 0;
        public virtual ResistenciaTipo ResistenciaTipo { get; set; }
        
        public float Mpa { get; set; }
        
        public int Consumo { get; set; }
        
        public int? PedraCodigo { get; set; } = 0;
        public virtual Pedra Pedra { get; set; }
        
        public int? SlumpCodigo { get; set; } = 0;
        public virtual Slump Slump { get; set; }

        public string SlumpNotaFiscal { get; set; }

        public int? UsoCodigo { get; set; } = 0;
        public virtual Uso Uso { get; set; }

        public string PecaConcretar { get; set; }

        public string Andar { get; set; }

        public int ObraTracoSequencia { get; set; }
        
        public float VolumeTotal { get; set; }
        
        public float VolumeEntregue { get; set; }
        
        public float VolumePorCarga { get; set; }
        
        public DateTime DataConcretagem { get; set; }
        
        public string Horario { get; set; }

        public string NecessitaConfirmacao { get; set; }

        public float VolumeLiberado { get; set; }

        public int IntervaloEmMinutosEntreCargas { get; set; }
        
        public int ObraBombaSequencia { get; set; }

        public int DistanciaTubulacao { get; set; }
        public float ValorAdicionalTubulacao { get; set; }

        public string HorarioBomba { get; set; }

        public string EquipamentoBombaCodigo { get; set; } = "";

        public int? TracoPesadoResistenciaTipoCodigo { get; set; } = 0;
        public virtual ResistenciaTipo TracoPesadoResistenciaTipo { get; set; }

        public float TracoPesadoMpa { get; set; }

        public int TracoPesadoConsumo { get; set; }

        public int? TracoPesadoPedraCodigo { get; set; } = 0;
        public virtual Pedra TracoPesadoPedra { get; set; }

        public int? TracoPesadoSlumpCodigo { get; set; } = 0;
        public virtual Slump TracoPesadoSlump { get; set; }

        public int? TracoPesadoUsoCodigo { get; set; } = 0;
        public virtual Uso TracoPesadoUso { get; set; }

        public int VibradorQuantidade { get; set; }

        public float VibradorValorUnitario { get; set; }

        public float VibradorValorTotal { get; set; }

        public int? VibradorVendedorCodigo { get; set; } = 0;
        public virtual Vendedor VibradorVendedor { get; set; }

        public EProgramacaoStatus Status { get; set; }

        public string Solicitante { get; set; }

        public string Observacao { get; set; }

        public bool TemNotaFicalEmitida { get; set; }

        public string FaxSimNao { get; set; }

        public virtual ICollection<ProgramacaoDemaisServicos> DemaisServicos { get; set; }

        public int ContinuidadeDaSequencia { get; set; }

        public string ObservacaoInterna { get; set; }

        public string DataHoraSolicitacao { get; set; }

        public string IdCadastro { get; set; }
        public string IdAtual { get; set; }

        public string CorpoDeProvaTipo { get; set; }
        public int CorpoDeProvaQuantidade { get; set; }
        public int CorpoDeProvaIntervalo { get; set; }
        public int CorpoDeProvaMoldador { get; set; }
        public string CorpoDeProvaMoldagemRemota { get; set; }

        public int ObraFrenteSequencia { get; set; }
        
        public decimal ValorTotal { get; set; }
        public decimal ValorConcreto { get; set; }
        public decimal ValorExtras { get; set; }
        public decimal ValorBomba { get; set; }
        public decimal ValorDemaisServicos { get; set; }
        public string ExternalId { get; set; }

        public int TempoBtNaObra { get; set; }

        public int TempoAteAObra { get; set; }

        public int TempoDescarga { get; set; }

        public decimal ValorTotalRemessasEmitidas { get; set; }
        public ECanceladoPor CanceladoPor { get; set; }

        public bool PossuiEndereco()
        {
           return EnderecoLogradouro == "" || (EnderecoNumero == 0 && EnderecoComplemento == "") || EnderecoCep == "";
        }

        public void setValorTotalRemessasEmitidas(decimal valorTotalRemessas)
        {
            ValorTotalRemessasEmitidas = valorTotalRemessas;
        }
    }
}
